src/generic/stage2_build_tape.h

```diff
@@ -113,9 +113,34 @@ class structural_iterator {
   uint8_t c{0};  // used to track the (structural) character we are looking at
 };
 
+struct number_writer {
```

Comment:
Why is number_writer its own struct? (This is a genuine question.)

Reply:
Oh, this is because of the `never_inline parse_large_integer()` function, which needs a pointer to something to write out to. Since we can't pass dom::parser anymore, because it doesn't have the write functions in it, we need something else. We can't pass structural_parser, though, because passing *that* to a never_inline function prevents the compiler from enregistering it and absolutely destroys performance. Instead, we pass a copy of this around, which is really just passing the doc_parser pointer around.